packages/next-swc/crates/next-api/src/dynamic_imports.rs

```diff
@@ -0,0 +1,395 @@
+use std::collections::HashMap;
+
+use anyhow::{bail, Result};
+use indexmap::IndexMap;
+use turbo_tasks::{
+    graph::{GraphTraversal, NonDeterministic},
+    Value, Vc,
+};
+use turbopack_binding::{
+    swc::core::ecma::{
+        ast::{CallExpr, Callee, Expr, Ident, Lit},
+        visit::{Visit, VisitWith},
+    },
+    turbo::tasks_fs::FileSystemPath,
+    turbopack::{
+        build::BuildChunkingContext,
+        core::{
+            chunk::{
+                availability_info::AvailabilityInfo, ChunkableModule, ChunkingContext,
+                EvaluatableAssets,
+            },
+            ident::AssetIdent,
+            issue::{Issue, IssueExt, IssueSeverity, OptionIssueSource},
+            module::Module,
+            output::OutputAssets,
+            reference::primary_referenced_modules,
+            reference_type::EcmaScriptModulesReferenceSubType,
+            resolve::{origin::PlainResolveOrigin, parse::Request, pattern::Pattern},
+        },
+        ecmascript::{
+            chunk::{EcmascriptChunkPlaceable, EcmascriptChunkingContext},
+            parse::ParseResult,
+            resolve::esm_resolve,
+            EcmascriptModuleAsset,
+        },
+    },
+};
+
+pub(crate) async fn collect_chunk_group(
+    chunking_context: Vc<BuildChunkingContext>,
+    dynamic_import_entries: IndexMap<Vc<Box<dyn Module>>, DynamicImportedModules>,
+    availability_info: Value<AvailabilityInfo>,
+) -> Result<Vc<DynamicImportedChunks>> {
+    let mut chunks_hash: HashMap<String, Vc<OutputAssets>> = HashMap::new();
+    let mut dynamic_import_chunks = IndexMap::new();
+
+    // Iterate over the collected import mappings, and create a chunk for each
+    // dynamic import.
+    for (origin_module, dynamic_imports) in dynamic_import_entries {
+        for (imported_raw_str, imported_module) in dynamic_imports {
+            let chunk = if let Some(chunk) = chunks_hash.get(&imported_raw_str) {
+                *chunk
+            } else {
+                let Some(imported_module) =
+                    Vc::try_resolve_sidecast::<Box<dyn EcmascriptChunkPlaceable>>(imported_module)
+                        .await?
+                else {
+                    bail!("module must be evaluatable");
+                };
+
+                // [Note]: this seems to create a duplicated chunk for the same module to the original import() call
+                // and the explicit chunk we ask in here. So there'll be at least 2
+                // chunks for the same module, relying on
+                // naive hash to have additonal
+                // chunks in case if there are same modules being imported in differnt
+                // origins.
+                let chunk =
+                    imported_module.as_chunk(Vc::upcast(chunking_context), availability_info);
```

Comment:
@jridgewell 

So it looks like we'd like to avoid to create chunk directly from the module (https://github.com/vercel/next.js/pull/56506); any recommendations for the workaround?

Reply:
trying to use as_chunk_item + as_chunk combination for now